### PR TITLE
(feat) Add PreCodec modifier

### PR DESCRIPTION
**Pull request by justinkaseman**

### CODE REVIEW

**Review by pablolagreca — DISMISSED** [review]

File: pkg/codec/config.go [hunk 1]
```diff
@@ -199,6 +203,33 @@ func (h *HardCodeModifierConfig) MarshalJSON() ([]byte, error) {
...
+type PreCodecModifierConfig struct {
+	// A map of a path of properties to encoding scheme.
+	// If the path leads to an array, encoding will occur on every entry.
+	//
+	// Example: "a.b" -> "uint256 Value"
+	Fields map[string]string
+	// A factory function that given an encoding scheme, returns a specific codec.
+	// This allows encoding and decoding implementations to be handled outside of the modifier.
+	// The map value given to "Fields" will be used to initialize the codec.
+	CodecFactory func(typeABI string) types.RemoteCodec
```
[the remark this answers is not in the document]
Comment (ilija42) [reply pkg/codec/config.go:219]: We actually have a generic address modifier which is injected in the relayer

File: pkg/codec/precodec.go [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+		modifierBase: modifierBase[string]{
+			fields:           fields,
+			onToOffChainType: map[reflect.Type]reflect.Type{},
+			offToOnChainType: map[reflect.Type]reflect.Type{},
+		},
+		codecFactory: codecFactory,
+		codecs:       make(map[string]types.RemoteCodec),
+	}
+
+	// set up a codec each unique ABI
```
Comment (ilija42) [inline pkg/codec/precodec.go:25]: abi is EVM leakage

File: pkg/codec/config.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (justinkaseman) [reply pkg/codec/config.go:219]: [Here](https://github.com/smartcontractkit/chainlink/pull/15538/files#diff-698f9f7fab7a3bafb8b43cdebb9407e15a7e1c9a7914e00c811dbab73dd0a94fR76) is an example of what one would look like for the current Capability Encoder, which currently only supports EVM.

Usage through the OCR3 Capability looks like passing it config as so:
```
consensus := ocr3cap.ReduceConsensusConfig[computeOutput]{
		Encoder: ocr3cap.EncoderEVM,
		EncoderConfig: map[string]any{
		    "subabi": map[string]any{"Reports.Bundle": "uint256 Ask, uint256 Bid"},
		    "abi": "(bytes32 FeedID, bytes Bundle, uint32 Timestamp)[] Reports",
		},
``` 

--

I'm not sure that I see a use case for passing more than one factory at once into the modifier. The codec gets injected one level up in the Encoder. Imagining a different chain Relayer (e.g. a non-EVM chain), the switch case would live in the Encoder and pick which factory to use in the modifier based on the chain family.

One fringe use case may be mixed encodings, such as wanting part of the report encoded for EVM and one part for non-EVM. For this we could just use multiple chained modifiers, one per chain family. So the user would pass something more like:
```
		    "subabi": map[string]any{
                           "evm": map[string]any{"Reports.BundleEVM": "uint256 Ask, uint256 Bid"},
                           "nonevm": map[string]any{"Reports.BundleNonEVM": "uint256 Ask, uint256 Bid"},
                     }
```

File: pkg/codec/precodec.go (hunk 2 above)
Comment (justinkaseman) [reply pkg/codec/precodec.go:25]: Good point! Renaming

File: pkg/codec/config.go [hunk 3]
```diff
@@ -199,6 +203,33 @@ func (h *HardCodeModifierConfig) MarshalJSON() ([]byte, error) {
 	})
 }
 
+// PreCodec creates a modifier that will transform data using a preliminary encoding/decoding step.
+// 'Off-chain' values will be overwritten with the encoded data as a byte array.
+// 'On-chain' values will be typed using the optimistic types from the codec.
+// This is useful when wanting to move the data as generic bytes.
+type PreCodecModifierConfig struct {
```
[the remark this answers is not in the document]
Comment (justinkaseman) [reply pkg/codec/config.go:246]: Added an example! Let me know what else you think is missing and I can add more words

File: pkg/codec/config.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (ilija42) [reply pkg/codec/config.go]: I was thinking of this https://github.com/smartcontractkit/chainlink/blob/9d8d9f48b92cab72de51e4484cc7b0833b1ccf6b/core/services/relay/evm/chain_reader.go#L145

[the remark this answers is not in the document]
Comment (ilija42) [reply pkg/codec/config.go]: If we can make this modifier more static then we wouldn't break the serialization

[the remark this answers is not in the document]
Comment (justinkaseman) [reply pkg/codec/config.go]: Pushed a change similar to AddressModifier!

File: pkg/codec/config.go [hunk 4]
```diff
@@ -199,6 +203,71 @@ func (h *HardCodeModifierConfig) MarshalJSON() ([]byte, error) {
...
+//
+//					type B struct {
+//						C string
+//						D string
+//					}
+//
+//				And the fields config defined as:
+//			 		{"A": "string C, string D"}
+//
+//				The codec factory function returns a RemoteCodec that provides an implementation for encoding/decoding
```
Comment (ilija42) [inline pkg/codec/config.go:226]: Does this comment need updating now?

**Review by ilija42 — COMMENTED** [review]
Looks better now, just need the accompanying EVM PR, and maybe adding a function in other repos(at least Solana) that says that this modifier is not supported

File: pkg/codec/config.go (hunk 4 above)
Comment (justinkaseman) [reply pkg/codec/config.go:226]: Yes, good catch

**Review by ilija42 — APPROVED** [review]
looks good now

**Review by cedric-cordenier — APPROVED** [review]

**Pull request merged**